File: src/euphony/lib/util/EuCodec.java
package euphony.lib.util;

import java.util.Arrays;

public class EuCodec {
	public static final int BASE40 = 40;
	public StringBuilder chars = new StringBuilder(BASE40);
	protected static byte[] base40Index = new byte[256];
	{
		chars.append('\0');
		for(char ch = 'a'; ch <= 'z'; ch++) chars.append(ch);
		for(char ch = '0'; ch <= '9'; ch++) chars.append(ch);
		chars.append("-:.");
		Arrays.fill(base40Index, (byte) -1);
		for(byte i = 0; i < chars.length(); i++)
			base40Index[chars.charAt(i)] = i;
	}
}

File: Euphony/src/main/kotlin/euphony/lib/util/EuCodec.kt
package euphony.lib.util;

import java.util.Arrays;

open class EuCodec {
	private val chars = StringBuilder(BASE40)

	init {
		chars.append('\u0000')
		run {
			var ch = 'a'
			while (ch <= 'z') {
				chars.append(ch)
				ch++
			}
		}
		var ch = '0'
		while (ch <= '9') {
			chars.append(ch)
			ch++
		}
		chars.append("-:.")
		Arrays.fill(base40Index, (-1).toByte())
		for (i in 0 until chars.length)
			base40Index[chars[i] as Int] = i as Byte
	}

	companion object {
		private val BASE40 = 40
		var base40Index = ByteArray(256)
	}
}